crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -86,166 +100,365 @@ impl TryFrom<PoolData> for PoolInfo {
+            pools,
+            block_number: registered_pools.fetched_block_number,
+            ..Default::default()
+        });
+
         Ok(Self {
-            pools_by_token_pair,
             graph_api,
-            cache: Default::default(),
-            max_age,
+            pools_by_token_pair,
+            pools_checkpoint,
         })
     }
 
-    async fn get_pools_and_update_cache(&self, pool_ids: &[H160]) -> Result<Vec<PoolData>> {
-        tracing::debug!("get_pools_and_update_cache");
-        let pools = self.graph_api.get_pools_with_ticks_by_ids(pool_ids).await?;
-        tracing::debug!("pools len: {}", pools.len());
-        let now = Instant::now();
-        let mut cache = self.cache.lock().unwrap();
-        for pool in &pools {
-            cache.insert(
-                pool.id,
-                CachedPool {
-                    pool: pool.clone(),
-                    updated_at: now,
-                    requested_at: now,
-                },
-            );
-        }
-        Ok(pools)
-    }
-
-    /// Returns cached pools and ids of outdated pools.
-    fn get_cached_pools(&self, token_pairs: &HashSet<TokenPair>) -> (Vec<PoolData>, Vec<H160>) {
-        tracing::debug!("UniswapV3PoolFetcher::get_cached_pools");
+    /// For a given list of token pairs, fetches the pools for the ones that exist in the checkpoint.
+    /// For the ones that don't exist, flag as missing and expect to exist after the next maintenance run.
+    fn get(&self, token_pairs: &HashSet<TokenPair>) -> (HashMap<H160, PoolData>, u64) {
         let mut pool_ids = token_pairs
             .iter()
             .filter_map(|pair| self.pools_by_token_pair.get(pair))
             .flatten()
             .peekable();
 
-        tracing::debug!("pool_ids: {:?}", pool_ids);
+        tracing::debug!("get checkpoint for pool_ids: {:?}", pool_ids);
 
         match pool_ids.peek() {
             Some(_) => {
-                let now = Instant::now();
-                let mut cache = self.cache.lock().unwrap();
-                pool_ids.partition_map(|pool_id| match cache.get_mut(pool_id) {
-                    Some(entry)
-                        if now.saturating_duration_since(entry.updated_at) < self.max_age =>
-                    {
-                        tracing::debug!("returning cached pool: {:?}", pool_id);
-                        entry.requested_at = now;
-                        Either::Left(entry.pool.clone())
-                    }
-                    _ => {
-                        tracing::debug!("returning outdated pool: {:?}", pool_id);
-                        Either::Right(pool_id)
-                    }
-                })
+                let mut pools_checkpoint = self.pools_checkpoint.lock().unwrap();
+                let (existing_pools, missing_pools): (HashMap<H160, PoolData>, Vec<H160>) =
+                    pool_ids.partition_map(|pool_id| match pools_checkpoint.pools.get(pool_id) {
```

Comment:
This is cool... didn't know this existed in std!

Reply:
I believe I stole it from one of the @MartinquaXD implementations of price estimator cache :)